components/style/rule_tree/mod.rs

```diff
@@ -0,0 +1,491 @@
+
+impl RuleTree {
+    pub fn new() -> Self {
+        RuleTree {
+            root: StrongRuleNode::new(Box::new(RuleNode::root())),
+        }
+    }
+
+    pub fn root(&self) -> StrongRuleNode {
+        self.root.clone()
+    }
+
+    fn dump<W: Write>(&self, writer: &mut W) {
+        let _ = writeln!(writer, " + RuleTree");
+        self.root.get().dump(writer, 0);
+    }
+
+    pub fn dump_stdout(&self) {
+        let mut stdout = io::stdout();
+        self.dump(&mut stdout);
+    }
+
+    pub fn insert_ordered_rules<'a, I>(&self, iter: I) -> StrongRuleNode
+        where I: Iterator<Item=(&'a Arc<StyleRule>, Importance)>
+    {
+        let root = self.root.clone();
+        let mut current = self.root.clone();
+        for (rule, importance) in iter {
+            current = current.ensure_child(&root, rule, importance);
+        }
+        current
+    }
+
+    unsafe fn pop_from_free_list(&self) -> Option<WeakRuleNode> {
+        debug_assert!(thread_state::get().is_layout() &&
+                      !thread_state::get().is_worker());
+
+        let current = self.root.get().next_free.load(Ordering::SeqCst);
+        if current == FREE_LIST_SENTINEL {
+            return None;
+        }
+
+        let current = WeakRuleNode { ptr: current };
+
+        let node = &*current.ptr();
+        let next = node.next_free.swap(ptr::null_mut(), Ordering::SeqCst);
+        self.root.get().next_free.store(next, Ordering::SeqCst);
+
+        debug!("Popping from free list: cur: {:?}, next: {:?}", current.ptr(), next);
+
+        Some(current)
+    }
+
+    pub unsafe fn gc(&self) {
+        while let Some(weak) = self.pop_from_free_list() {
+            let needs_drop = {
+                let node = &*weak.ptr();
+                if node.refcount.load(Ordering::SeqCst) == 0 {
+                    node.remove_from_child_list();
+                    true
+                } else {
+                    false
+                }
+            };
+
+            debug!("GC'ing {:?}: {}", weak.ptr(), needs_drop);
+            if needs_drop {
+                let _ = Box::from_raw(weak.ptr());
+            }
+        }
+
+        debug_assert!(self.root.get().next_free.load(Ordering::SeqCst) == FREE_LIST_SENTINEL);
+    }
+}
+
+struct RuleNode {
+    /// The root node. Only the root has no root pointer, for obvious reasons.
+    root: Option<StrongRuleNode>,
+
+    /// The parent rule node. Only the root has no parent.
+    parent: Option<StrongRuleNode>,
+
+    /// The actual style rule that matched. Only the root has no declarations.
+    rule: Option<Arc<StyleRule>>,
+
+    /// The importance of the declarations relevant in the style rule,
+    /// meaningless in the root node.
+    importance: Importance,
+
+    children: RuleChildrenList,
+    refcount: AtomicUsize,
+    next: AtomicPtr<RuleNode>,
+    prev: AtomicPtr<RuleNode>,
+
+    /// The next item in the rule tree free list, that starts on the root node.
+    next_free: AtomicPtr<RuleNode>,
+}
```

Comment:
Yeah, I thought about making it generic, but that might get tricky. I think I can write a few unit tests that stress this a bit, or something.

And yeah, totally agree in the documentation part, this is ready for review, but needs docs before landing.